Pegasus/include/Math.hpp

```diff
@@ -30,13 +659,450 @@ class JacobiEigenvalue
+                            && visibleFacesSet.find(&*adjFace) == visibleFacesSet.end())
+                        {
+                            //If face is visible add it to the visible faces set
+                            if (adjFace->GetHyperPlane().SignedDistance(*extremalVertex) > 0)
+                            {
+                                visibleFaces.push_back(adjFace);
+                                visibleFacesSet.insert(&*adjFace);
+
+                                std::vector<typename Vertices::iterator>& faceVertices = adjFace->GetVertices();
+                                markedVertexIterators.insert(markedVertexIterators.end(), faceVertices.begin(), faceVertices.end());
+                                std::vector<typename Vertices::iterator>().swap(faceVertices);
+                            }
+                            //If face is not visible then find a ridge and save it
+                            else
+                            {
+                                std::vector<size_t> ridgeIndices;
+                                ridgeIndices.reserve(2);
+
+                                auto adjFaceIndices = adjFace->GetIndices();
+                                auto faceIndices = (*visibleFaceIt)->GetIndices();
+                                std::set_intersection(adjFaceIndices.begin(), adjFaceIndices.end(),
+                                    faceIndices.begin(), faceIndices.end(), std::back_inserter(ridgeIndices));
+
+                                horizonRidges.emplace_back(ridgeIndices[0], ridgeIndices[1]);
+                            }
+                        }
+                    }
+                    while (++adjHedsFaceIt != adjHedsFaceBegin);
+                }
+
+                m_convexHullVertices.push_back(extremalVertex);
+
+                //Remove visible faces
+                for (typename Faces::iterator visibleFaceIt : visibleFaces)
+                {
+                    faceStack.erase(std::remove_if(faceStack.begin(), faceStack.end(),
+                        [&visibleFaceIt](typename Faces::iterator& it)
+                    {
+                        return &*it == &*visibleFaceIt;
+                    }), faceStack.end());
+                    RemoveFace(visibleFaceIt);
+                }
+
+                //Make new faces from horizon ridges and an extremal point
+                for (auto& ridge : horizonRidges)
+                {
+                    faceStack.push_back(
+                        MakeFace(ridge.first, ridge.second, extremalVertexIndex, markedVertexIterators)
+                    );
+                }
+            }
+        }
+    }
+
+    /**
+     * @brief Inserts a new face into current convex hull.
+     * @param vertexIndex1 index of a face
+     * @param vertexIndex2 index of a face
+     * @param vertexIndex3 index of a face
+     * @param vertices modifiable vertex buffer
```

Comment:
if you have arguments that will be changed, mark them accordingly: `@param[in]` `@param[out]` `@param[in,out]`